pkg/tls/reloader_test.go

```diff
@@ -0,0 +1,246 @@
+			CommonName: commonName,
+		},
+		DNSNames:              []string{commonName},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().Add(time.Hour * 24),
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	certDER, err := x509.CreateCertificate(rand.Reader, &template, &template, &privateKey.PublicKey, privateKey)
+	require.NoError(t, err)
+
+	certPEM = pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certDER,
+	})
+
+	keyPEM = pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	})
+
+	return certPEM, keyPEM
+}
+
+func TestReloader_CertificateRotation(t *testing.T) {
+	tempDir, err := os.MkdirTemp("", "tls-test-")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempDir)
+
+	certFile := filepath.Join(tempDir, "cert.pem")
+	keyFile := filepath.Join(tempDir, "key.pem")
+
+	certPEM1, keyPEM1 := generateSelfSignedCert(t, "test1.local")
+	err = os.WriteFile(certFile, certPEM1, 0o600)
+	require.NoError(t, err)
+	err = os.WriteFile(keyFile, keyPEM1, 0o600)
+	require.NoError(t, err)
+
+	log := logger.GetLogger("tls-test")
+	reloader, err := NewReloader(certFile, keyFile, log)
+	require.NoError(t, err)
+
+	defer reloader.Stop()
+
+	initialCert, err := reloader.getCertificate(nil)
+	require.NoError(t, err)
+	leafCert, err := x509.ParseCertificate(initialCert.Certificate[0])
+	require.NoError(t, err)
+	assert.Equal(t, "test1.local", leafCert.Subject.CommonName)
+
+	certPEM2, keyPEM2 := generateSelfSignedCert(t, "test2.local")
+	err = os.WriteFile(certFile, certPEM2, 0o600)
+	require.NoError(t, err)
+	err = os.WriteFile(keyFile, keyPEM2, 0o600)
+	require.NoError(t, err)
+
+	err = reloader.reloadCertificate()
+	require.NoError(t, err)
+
+	updatedCert, err := reloader.getCertificate(nil)
+	require.NoError(t, err)
+	leafCert, err = x509.ParseCertificate(updatedCert.Certificate[0])
+	require.NoError(t, err)
+	assert.Equal(t, "test2.local", leafCert.Subject.CommonName)
+
+	tlsConfig := reloader.GetTLSConfig()
+	configCert, err := tlsConfig.GetCertificate(nil)
+	require.NoError(t, err)
+	leafCert, err = x509.ParseCertificate(configCert.Certificate[0])
+	require.NoError(t, err)
+	assert.Equal(t, "test2.local", leafCert.Subject.CommonName)
+}
+
+func TestReloader_FileOperations(t *testing.T) {
+	tempDir := t.TempDir()
+	certFile := filepath.Join(tempDir, "cert.pem")
+	keyFile := filepath.Join(tempDir, "key.pem")
+
+	// Create initial files
+	certPEM, keyPEM := generateSelfSignedCert(t, "initial.local")
+	require.NoError(t, os.WriteFile(certFile, certPEM, 0o600))
+	require.NoError(t, os.WriteFile(keyFile, keyPEM, 0o600))
+
+	// Create reloader
+	log := logger.GetLogger("tls-test")
+	reloader, err := NewReloader(certFile, keyFile, log)
+	require.NoError(t, err)
+	require.NoError(t, reloader.Start())
+	defer reloader.Stop()
```

Comment:
This section should be included in each `t.Run` to prevent accidental changes by subcases.